File: scripts/scrape-roadmap.js
// scripts/scrape-roadmap.js
// Node.js script to scrape resources from roadmap.sh
// Run with: node scripts/scrape-roadmap.js

import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// Since we can't use Puppeteer in this basic setup, we'll use a proxy approach
// or fetch from roadmap.sh API if available, or use a service
// For now, we'll create a structure that can be populated

/**
 * Roadmap.sh role to URL mapping
 */
const roadmapRoles = {
  'junior-web-dev': {
    name: 'Junior Web Developer',
    url: 'https://roadmap.sh/frontend', // Junior web dev typically follows frontend path
    skills: ['html', 'css', 'javascript', 'react', 'git', 'github']
  },
  'data-analyst': {
    name: 'Data Analyst',
    url: 'https://roadmap.sh/data-analyst',
    skills: ['python', 'sql', 'excel', 'pandas', 'tableau']
  },
  'backend-dev': {
    name: 'Backend Developer',
    url: 'https://roadmap.sh/backend',
    skills: ['python', 'javascript', 'nodejs', 'express', 'sql', 'mongodb']
  },
  'fullstack-dev': {
    name: 'Full Stack Developer',
    url: 'https://roadmap.sh/full-stack',
    skills: ['html', 'css', 'javascript', 'react', 'nodejs', 'express', 'sql']
  },
  'python-dev': {
    name: 'Python Developer',
    url: 'https://roadmap.sh/python',
    skills: ['python', 'django', 'flask', 'sql']
  },
  'frontend-dev': {
    name: 'Frontend Developer',
    url: 'https://roadmap.sh/frontend',
    skills: ['html', 'css', 'javascript', 'react', 'typescript']
  },
  'devops-engineer': {
    name: 'DevOps Engineer',
    url: 'https://roadmap.sh/devops',
    skills: ['docker', 'aws', 'git', 'github']
  },
  'database-admin': {
    name: 'Database Administrator',
    url: 'https://roadmap.sh/postgresql-dba',
    skills: ['sql', 'postgresql', 'mysql', 'mongodb']
  }
};

/**
 * Skill-specific roadmap URLs
 */
const skillRoadmaps = {
  'html': 'https://roadmap.sh/frontend',
  'css': 'https://roadmap.sh/frontend',
  'javascript': 'https://roadmap.sh/frontend',
  'typescript': 'https://roadmap.sh/frontend',
  'react': 'https://roadmap.sh/react',
  'nodejs': 'https://roadmap.sh/backend',
  'express': 'https://roadmap.sh/backend',
  'python': 'https://roadmap.sh/python',
  'django': 'https://roadmap.sh/python',
  'flask': 'https://roadmap.sh/python',
  'java': 'https://roadmap.sh/java',
  'sql': 'https://roadmap.sh/postgresql-dba',
  'mongodb': 'https://roadmap.sh/mongodb',
  'postgresql': 'https://roadmap.sh/postgresql-dba',
  'mysql': 'https://roadmap.sh/postgresql-dba',
  'docker': 'https://roadmap.sh/devops',
  'aws': 'https://roadmap.sh/aws',
  'git': 'https://roadmap.sh/frontend',
  'github': 'https://roadmap.sh/frontend',
  'pandas': 'https://roadmap.sh/data-analyst',
  'tableau': 'https://roadmap.sh/data-analyst',
  'excel': 'https://roadmap.sh/data-analyst'
};

/**
 * Extract resources from roadmap.sh page
 * Note: This is a template. You'll need to use Puppeteer or a similar tool
 * to actually scrape the content since roadmap.sh is likely a React app
 */
async function scrapeRoadmapResources(skillId, roadmapUrl) {
  // This is a placeholder function
  // In production, you would use Puppeteer to scrape the page
  // For now, we return the roadmap URL as a resource
  
  console.log(`Scraping resources for ${skillId} from ${roadmapUrl}`);
  
  // Placeholder: In real implementation, you would:
  // 1. Use Puppeteer to load the page
  // 2. Wait for React to render
  // 3. Extract resource links (YouTube, articles, etc.)
  // 4. Return structured data
  
  return {
    skillId,
    roadmapUrl,
    resources: [
      {
        title: `${skillId} Roadmap - Complete Learning Path`,
        url: roadmapUrl,
        platform: 'roadmap.sh',
        type: 'interactive',
        description: 'Complete roadmap with free resources'
      }
    ]
  };
}

/**
 * Main function to scrape all resources
 */
async function main() {
  console.log('Starting roadmap.sh resource scraping...');
  console.log('Note: This script requires Puppeteer for actual scraping.');
  console.log('For now, it generates the structure. See scripts/scrape-roadmap-puppeteer.js for full implementation.\n');
  
  const allResources = [];
  
  // Process each skill
  for (const [skillId, roadmapUrl] of Object.entries(skillRoadmaps)) {
    const result = await scrapeRoadmapResources(skillId, roadmapUrl);
    allResources.push(...result.resources.map(resource => ({
      id: `${skillId}-roadmap-${Date.now()}`,
      skillId: skillId,
      ...resource
    })));
  }
  
  // Save to file
  const outputPath = path.join(__dirname, '../src/data/roadmap-resources-scraped.json');
  fs.writeFileSync(outputPath, JSON.stringify(allResources, null, 2));
  
  console.log(`\nScraped ${allResources.length} resources`);
  console.log(`Saved to: ${outputPath}`);
  console.log('\nNext steps:');
  console.log('1. Install Puppeteer: npm install puppeteer');
  console.log('2. Use the Puppeteer version: scripts/scrape-roadmap-puppeteer.js');
  console.log('3. Merge results into resources.json');
}

// Run main function
main().catch(console.error);

export { scrapeRoadmapResources, skillRoadmaps, roadmapRoles };

File: scripts/scrape-roadmap.py
#!/usr/bin/env python3
"""
Python script to scrape resources from roadmap.sh
Install dependencies: pip install requests beautifulsoup4 selenium
Run: python scripts/scrape-roadmap.py
"""

import requests
from bs4 import BeautifulSoup
import json
import time
import os
from urllib.parse import urljoin, urlparse

# Roadmap.sh base URL
BASE_URL = "https://roadmap.sh"

# Skill to roadmap URL mapping
SKILL_ROADMAPS = {
    'html': 'https://roadmap.sh/frontend',
    'css': 'https://roadmap.sh/frontend',
    'javascript': 'https://roadmap.sh/frontend',
    'typescript': 'https://roadmap.sh/frontend',
    'react': 'https://roadmap.sh/react',
    'nodejs': 'https://roadmap.sh/backend',
    'express': 'https://roadmap.sh/backend',
    'python': 'https://roadmap.sh/python',
    'django': 'https://roadmap.sh/python',
    'flask': 'https://roadmap.sh/python',
    'java': 'https://roadmap.sh/java',
    'sql': 'https://roadmap.sh/postgresql-dba',
    'mongodb': 'https://roadmap.sh/mongodb',
    'postgresql': 'https://roadmap.sh/postgresql-dba',
    'mysql': 'https://roadmap.sh/postgresql-dba',
    'docker': 'https://roadmap.sh/devops',
    'aws': 'https://roadmap.sh/aws',
    'git': 'https://roadmap.sh/frontend',
    'github': 'https://roadmap.sh/frontend',
    'pandas': 'https://roadmap.sh/data-analyst',
    'tableau': 'https://roadmap.sh/data-analyst',
    'excel': 'https://roadmap.sh/data-analyst'
}

def determine_resource_type(url, platform):
    """Determine resource type based on URL and platform"""
    url_lower = url.lower()
    
    if 'youtube.com' in url_lower or 'youtu.be' in url_lower:
        return 'video'
    elif 'freecodecamp.org' in url_lower or 'codecademy.com' in url_lower:
        return 'interactive'
    elif 'coursera.org' in url_lower or 'udemy.com' in url_lower or 'edx.org' in url_lower:
        return 'course'
    elif 'medium.com' in url_lower or 'dev.to' in url_lower:
        return 'article'
    else:
        return 'docs'

def determine_platform(url):
    """Determine platform from URL"""
    url_lower = url.lower()
    domain = urlparse(url).netloc.lower()
    
    if 'youtube.com' in url_lower or 'youtu.be' in url_lower:
        return 'YouTube'
    elif 'freecodecamp.org' in url_lower:
        return 'freeCodeCamp'
    elif 'codecademy.com' in url_lower:
        return 'Codecademy'
    elif 'coursera.org' in url_lower:
        return 'Coursera'
    elif 'udemy.com' in url_lower:
        return 'Udemy'
    elif 'edx.org' in url_lower:
        return 'edX'
    elif 'mdn.io' in url_lower or 'mozilla.org' in url_lower:
        return 'MDN'
    elif 'w3schools.com' in url_lower:
        return 'W3Schools'
    elif 'medium.com' in url_lower:
        return 'Medium'
    elif 'dev.to' in url_lower:
        return 'Dev.to'
    elif 'github.com' in url_lower:
        return 'GitHub'
    elif 'khanacademy.org' in url_lower:
        return 'Khan Academy'
    else:
        return domain.split('.')[-2].title() if '.' in domain else 'Unknown'

def scrape_roadmap_page(skill_id, roadmap_url):
    """Scrape resources from a roadmap.sh page"""
    print(f"Scraping {skill_id} from {roadmap_url}")
    
    try:
        headers = {
            'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36'
        }
        
        response = requests.get(roadmap_url, headers=headers, timeout=30)
        response.raise_for_status()
        
        soup = BeautifulSoup(response.content, 'html.parser')
        resources = []
        
        # Find all links on the page
        links = soup.find_all('a', href=True)
        
        # Resource platforms to look for
        resource_domains = [
            'youtube.com', 'youtu.be',
            'freecodecamp.org', 'codecademy.com',
            'coursera.org', 'udemy.com', 'edx.org',
            'mdn.io', 'mozilla.org', 'w3schools.com',
            'medium.com', 'dev.to',
            'github.com', 'khanacademy.org'
        ]
        
        seen_urls = set()
        
        for link in links:
            href = link.get('href', '').strip()
            text = link.get_text(strip=True)
            
            if not href or not text or len(text) < 5:
                continue
            
            # Convert relative URLs to absolute
            if href.startswith('/'):
                href = urljoin(BASE_URL, href)
            elif not href.startswith('http'):
                continue
            
            # Check if it's a resource link
            is_resource = any(domain in href.lower() for domain in resource_domains)
            
            if is_resource and href not in seen_urls:
                seen_urls.add(href)
                
                platform = determine_platform(href)
                resource_type = determine_resource_type(href, platform)
                
                resources.append({
                    'skillId': skill_id,
                    'title': text[:200],  # Limit title length
                    'url': href,
                    'platform': platform,
                    'type': resource_type
                })
        
        print(f"  Found {len(resources)} resources")
        return resources
        
    except Exception as e:
        print(f"  Error scraping {roadmap_url}: {str(e)}")
        return []

def main():
    """Main scraping function"""
    print("Starting roadmap.sh resource scraping...")
    print("This may take a few minutes...\n")
    
    all_resources = []
    processed_urls = set()
    
    for skill_id, roadmap_url in SKILL_ROADMAPS.items():
        if roadmap_url in processed_urls:
            print(f"\nSkipping {skill_id} - URL already processed")
            continue
        
        resources = scrape_roadmap_page(skill_id, roadmap_url)
        all_resources.extend(resources)
        processed_urls.add(roadmap_url)
        
        # Be respectful - wait between requests
        time.sleep(2)
    
    # Add unique IDs
    import time as time_module
    timestamp = int(time_module.time() * 1000)
    resources_with_ids = [
        {
            'id': f"{r['skillId']}-scraped-{timestamp}-{i}",
            **r
        }
        for i, r in enumerate(all_resources)
    ]
    
    # Save to file
    output_path = os.path.join(os.path.dirname(__file__), '../src/data/roadmap-resources-scraped.json')
    with open(output_path, 'w', encoding='utf-8') as f:
        json.dump(resources_with_ids, f, indent=2, ensure_ascii=False)
    
    print(f"\n✅ Scraping complete!")
    print(f"   Total resources found: {len(resources_with_ids)}")
    print(f"   Saved to: {output_path}")
    print(f"\nNote: If roadmap.sh uses React/JS rendering, you may need Selenium.")
    print(f"Next: Run 'node scripts/merge-resources.js' to merge into resources.json")

if __name__ == '__main__':
    main()

